File: src/geek time/c#/week03/18. 4Sum.cs
using System;
using System.Collections.Generic;

namespace week03 {
    public partial class Solution {
        public IList<IList<int>> FourSum(int[] nums, int target) {
            var list = new List<IList<int>>();
            if (nums == null || nums.Length < 4) {
                return list;
            }
            var a = 0;
            Array.Sort(nums);
            while (a < nums.Length - 3) {
                while (a > 0  && a < nums.Length - 1 && nums[a] == nums[a-1]) {
                    a++;
                }
                for (var b = a + 1; b < nums.Length - 2; b++) {
                    while (b > a + 1 && b < nums.Length - 2 && nums[b] == nums[b-1]) {
                        b++;
                    }
                    var c = b + 1;
                    var d = nums.Length - 1;
                    while (c < d) {
                        var sum = nums[a] + nums[b] + nums[c] + nums[d];
                        if (sum < target) {
                            while (c < d && nums[c++] == nums[c]);
                            continue;
                        }
                        if (sum > target) {
                            while (c < d && nums[d--] == nums[d]);
                            continue;
                        }
                        if (sum == target) {
                            list.Add(new List<int>() { nums[a], nums[b], nums[c], nums[d]});
                            while (c < d && nums[c++] == nums[c]);
                            while (c < d && nums[d--] == nums[d]);
                        }
                    }
                }
                a++;
            }
            return list;
        }
    }
}

File: src/geek time/typescript/week03/18. 4Sum.ts
function fourSum(nums: number[], target: number): number[][] {
    if (!nums || nums.length < 4) return [];
    const result = [];
    nums = nums.sort((a, b) => a - b);
    for (let i = 0; i < nums.length - 3; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;
        for (let j = i + 1; j < nums.length - 2; j++) {
            if (j > i + 1 && nums[j] === nums[j - 1]) continue;
            let a = j + 1;
            let b = nums.length - 1;
            while (a < b) {
                const sum = nums[i] + nums[j] + nums[a] + nums[b];
                if (sum < target) {
                    while (a < b && nums[a++] === nums[a]);
                } else if (sum > target) {
                    while (a < b && nums[b--] === nums[b]);
                } else {
                    result.push([nums[i], nums[j], nums[a], nums[b]]);
                    while (a < b && nums[a++] === nums[a]);
                    while (a < b && nums[b--] === nums[b]);
                }
            }
        }
    }
    return result;
}
